File: 2014-06-20-FP/Mangoes/m.accepted.scala
object Solution {
    def sat(mx: BigInt, m: Int, xs: List[(BigInt, BigInt)]) =
        ((xs map (x => x._1 + (m - 1) * x._2)) take m).sorted.sum <= mx

    @annotation.tailrec
    def solve(mx: BigInt, a: Int, b: Int, xs: List[(BigInt, BigInt)]): Int =
        if (a == b)
            a
        else {
            val m: Int = (a + b + 1) / 2
            if (sat(mx, m, xs))
                solve(mx, m, b, xs)
            else
                solve(mx, a, m - 1, xs)
        }

    def main(args: Array[String]) = {
        val ps = readLine.split(" ").map(x => BigInt(x))
        val as = readLine.split(" ").map(x => BigInt(x))
        val hs = readLine.split(" ").map(x => BigInt(x))
        val xs = (as zip hs) sortWith ((x, y) => (x._2 < y._2) || (x._2 == y._2 && x._1 < y._1))
        println(solve(ps(1), 0, ps(0).toInt, xs.toList))
    }
}

File: 2014-06-20-FP/Mangoes/m.accepted.fs
open System

let readIntList _ =
    let str = Console.ReadLine ()
    List.map (fun x -> int64 x) (List.ofSeq (str.Split [|' '|]))

let rec take n xs =
    if n = 0L || (List.isEmpty xs)
        then []
        else (List.head xs) :: (take (n - 1L) (List.tail xs))

let sat mx m xs =
    List.sum (take m (List.sort (List.map (fun (w, k) -> w + (m - 1L) * k) xs))) <= mx

let rec solve mx a b xs =
    if a = b
        then a
        else
            let m = (a + b + 1L) / 2L
            if sat mx m xs
                then solve mx m b xs
                else solve mx a (m - 1L) xs

let main _ =
    let ps = readIntList ()
    let ys = readIntList ()
    let hs = readIntList ()
    let xs = List.sortWith (fun (a, x) (b, y) -> compare x y) (List.zip ys hs)
    Console.WriteLine("{0}", solve (ps.Item(1)) 0L (ps.Item(0)) xs)

main ()
